cg/apps/demultiplex/flowcell.py

```diff
@@ -0,0 +1,5083 @@
+import logging
+from pathlib import Path
+
+from cg.models.demultiplex.run_parameters import RunParameters
+from cgmodels.demultiplex.sample_sheet import get_sample_sheet_from_file
+from cgmodels.exceptions import SampleSheetError
+from pydantic import ValidationError
+
+LOG = logging.getLogger(__name__)
+
+
+class Flowcell:
+    """Class to collect information about flowcell directories and there particular files"""
+
+    def __init__(self, flowcell_path: Path):
+        LOG.info("Instantiating Flowcell with path %s", flowcell_path)
+        self.path = flowcell_path
+        LOG.info("Set flowcell id to %s", self.flowcell_id)
+
+    @property
+    def flowcell_id(self) -> str:
+        base_name: str = self.path.name.split("_")[-1]
+        return base_name[1:]
+
+    @property
+    def sample_sheet_path(self) -> Path:
+        return Path(self.path, "SampleSheet.csv")
+
+    def sample_sheet_exists(self) -> bool:
+        """Check if sample sheet exists"""
+        return self.sample_sheet_path.exists()
+
+    def validate_sample_sheet(self) -> bool:
+        """Validate if sample sheet is on correct format"""
+        try:
+            get_sample_sheet_from_file(infile=self.sample_sheet_path, sheet_type="S4")
+        except (SampleSheetError, ValidationError) as error:
+            LOG.warning("Invalid sample sheet")
+            LOG.warning(error)
+            return False
+        return True
+
+    @property
+    def run_parameters_path(self) -> Path:
+        return self.path / "runParameters.xml"
+
+    @property
+    def run_parameters_object(self) -> RunParameters:
+        if not self.run_parameters_path.exists():
+            raise FileNotFoundError(
+                f"Could not find run parameters file {self.run_parameters_path}"
+            )
+        return RunParameters(run_parameters_path=self.run_parameters_path)
+
+    def is_sequencing_done(self) -> bool:
+        """Check if sequencing is done
+
+        This is indicated by that the file RTAComplete.txt exists
+        """
+        return Path(self.path, "RTAComplete.txt").exists()
+
+    def is_copy_completed(self) -> bool:
+        """Check if copy of flowcell is done
+
+        This is indicated by that the file CopyComplete.txt exists
+        """
+        return Path(self.path, "CopyComplete.txt").exists()
+
+    def is_demultiplexing_ongoing(self) -> bool:
+        """Check if demultiplexing is ongoing
+
+        This is indicated by if the file demuxstarted.txt exists (?)
```

Comment:
Demultiplexing is ongoing if demuxstarted.txt exists in the in_dir and demuxcomplete.txt does NOT exist in the out_dir.
If both exists demuxing has started and completed.

Reply:
Good point!!

Reply:
Note that demuxstarted.txt is currently created by checkfornewrun.bash You also need to create this file when starting demulitplexing from cg, or find a more elegant way of handling the demux status. The file is not generated by bcl2fastq or something.

Reply:
Thanks I will do it in the same way that it is done for spring compression

Reply:
same goes for the demuxcomplete.txt file btw, it's generated by our bash scripts and not bcl2fastq

Reply:
Updated this, thanks for the remark @barrystokman !